File: go/leetcode/addtwonumbers.go
package leetcode

func (ad *AddTwoNumbers) Solver(l1 *ListNode, l2 *ListNode) *ListNode {
	l3 := &ListNode{}
	l3ptr := l3

	var carry int

	for {
		var n1 int
		var n2 int

		if l1 != nil {
			n1 = l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			n2 = l2.Val
			l2 = l2.Next
		}

		n3 := n1 + n2 + carry

		if n3 >= 10 {
			carry = n3 / 10
			l3ptr.Val = n3 - 10
		} else {
			carry = 0
			l3ptr.Val = n3
		}

		if l1 != nil || l2 != nil || carry != 0 {
			l3ptr.Next = &ListNode{}
			l3ptr = l3ptr.Next
		} else {
			break
		}
	}

	return l3
}

File: typescript/src/leetcode/addtwonumbers.ts
/* eslint-disable no-param-reassign */
import { ListNode } from './types.js';

export default class AddTwoNumbers {
  public static Solver(l1?: ListNode, l2?: ListNode): ListNode | null {
    const l3 = new ListNode();
    let tmp = l3;

    while (l1 != null || l2 != null) {
      if (l1 != null) {
        tmp.val += l1.val;
        l1 = l1.next;
      }

      if (l2 != null) {
        tmp.val += l2.val;
        l2 = l2.next;
      }

      if (tmp.val > 9) {
        tmp.val -= 10;
        tmp.next = new ListNode(1);
      } else if (l1 != null || l2 != null) {
        tmp.next = new ListNode();
      }

      tmp = tmp.next;
    }

    return l3;
  }

  public static CleanSolver(l1?: ListNode, l2?: ListNode, carry: number = 0): ListNode | null {
    if (!l1 && !l2 && !carry) return null;

    const total: number = (l1 ? l1.val : 0) + (l2 ? l2.val : 0) + (carry || 0);
    carry = total > 9 ? 1 : 0;

    return new ListNode(total % 10, this.CleanSolver(l1?.next, l2?.next, carry));
  }
}
